rcl_action/test/rcl_action/test_action_server.cpp

```diff
@@ -0,0 +1,546 @@
+{
+  // Request to cancel all goals
+  rcl_action_cancel_request_t cancel_request = rcl_action_get_zero_initialized_cancel_request();
+  cancel_request.goal_info.stamp.sec = 0;
+  cancel_request.goal_info.stamp.nanosec = 0u;
+  rcl_action_cancel_response_t cancel_response = rcl_action_get_zero_initialized_cancel_response();
+  rcl_ret_t ret = rcl_action_process_cancel_request(
+    &this->action_server, &cancel_request, &cancel_response);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  EXPECT_NE(cancel_response.msg.goals_canceling.data, nullptr);
+  ASSERT_EQ(cancel_response.msg.goals_canceling.size, (size_t)NUM_GOALS);
+  rcl_action_goal_info_t * goal_info_out;
+  for (int i = 0; i < NUM_GOALS; ++i) {
+    goal_info_out = &cancel_response.msg.goals_canceling.data[i];
+    for (int j = 0; j < UUID_SIZE; ++j) {
+      EXPECT_EQ(goal_info_out->uuid[j], static_cast<uint8_t>(i + j));
+    }
+  }
+}
+
+TEST_F(TestActionServerCancelPolicy, test_action_process_cancel_request_single_goal)
+{
+  // Request to cancel a specific goal
+  rcl_action_cancel_request_t cancel_request = rcl_action_get_zero_initialized_cancel_request();
+  for (int i = 0; i < UUID_SIZE; ++i) {
+    cancel_request.goal_info.uuid[i] = static_cast<uint8_t>(i + 2);
+  }
+  rcl_action_cancel_response_t cancel_response = rcl_action_get_zero_initialized_cancel_response();
+  rcl_ret_t ret = rcl_action_process_cancel_request(
+    &this->action_server, &cancel_request, &cancel_response);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  EXPECT_NE(cancel_response.msg.goals_canceling.data, nullptr);
+  ASSERT_EQ(cancel_response.msg.goals_canceling.size, 1u);
+  rcl_action_goal_info_t * goal_info = &cancel_response.msg.goals_canceling.data[0];
+  for (int i = 0; i < UUID_SIZE; ++i) {
+    EXPECT_EQ(goal_info->uuid[i], static_cast<uint8_t>(i + 2));
+  }
+}
+
+TEST_F(TestActionServerCancelPolicy, test_action_process_cancel_request_by_time)
+{
+  // Request to cancel all goals at and before a specific time
+  const size_t time_index = 7;
+  rcl_action_cancel_request_t cancel_request = rcl_action_get_zero_initialized_cancel_request();
+  cancel_request.goal_info = this->goal_infos_out[time_index];
+  rcl_action_cancel_response_t cancel_response = rcl_action_get_zero_initialized_cancel_response();
+  rcl_ret_t ret = rcl_action_process_cancel_request(
+    &this->action_server, &cancel_request, &cancel_response);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  EXPECT_NE(cancel_response.msg.goals_canceling.data, nullptr);
+  ASSERT_EQ(cancel_response.msg.goals_canceling.size, time_index + 1);  // goals at indices [0, 7]
+  rcl_action_goal_info_t * goal_info_out;
+  for (size_t i = 0; i < cancel_response.msg.goals_canceling.size; ++i) {
+    goal_info_out = &cancel_response.msg.goals_canceling.data[i];
+    for (size_t j = 0; j < UUID_SIZE; ++j) {
+      EXPECT_EQ(goal_info_out->uuid[j], static_cast<uint8_t>(i + j));
+    }
```

Comment:
@jacobperron same as above but using `goals_info_out` instead.

Reply:
In this case, we would need to reconstruct the uuid in a similar loop before comparing anyways.